### PR TITLE
Use `map[string]interface{}` for `SampleProviderSpec` and `SampleIgnition`

### LINKED ISSUE #95: Define sample `Provider` spec as non string
# Summary

Currently the sample `ProviderSpec` in the `internal` package https://github.com/onmetal/machine-controller-manager-provider-onmetal/blob/main/pkg/internal/providerspec.go is defined as a string byte slice. We should rename it to `SampleProviderSpec` and use a `map[string]interface{}`instead. When used in the test cases we should then marshal the object into a JSON byte slice wherever it is used.

### LINKED ISSUE #96: Move `SampleIgnition` to `internal` pacakge
# Summary

Move the `SampleIgnition` in the create machine method (https://github.com/onmetal/machine-controller-manager-provider-onmetal/blob/main/pkg/onmetal/create_machine_test.go#L40) and move it to the internal package.

Also use here a `map[string]interface{}` and serialize it whenever it is needed. This will keep the code clean and more readable. Also extending the `SampleIgnition` will be more intuitive as we don't have to render it first via `butane` or fiddle around in the `[]byte` slice.

### PR DESCRIPTION
**Pull request by sujeet01**
# Proposed Changes

- Moved sample `Ignition` and `ProviderSpec` to `testdata` folder
- Changed sample `Ignition` and `ProviderSpec` string type to `map[string]interface{}`
- Corrected tests

Fixes #95 
Fixes #96

### CODE REVIEW

File: pkg/onmetal/create_machine_test.go [hunk 1]
```diff
@@ -141,12 +137,16 @@ var _ = Describe("CreateMachine", func() {
 				Name:      fmt.Sprintf("%s-ignition", machineName),
 			},
 		}
+
 		Eventually(Object(ignition)).Should(SatisfyAll(
-			HaveField("Data", map[string][]byte{
-				"ignition.json": SampleIgnition,
-			}),
+			HaveField("Data", HaveKey("ignition.json")),
```
Comment (afritzler) [inline pkg/onmetal/create_machine_test.go:142]: Use `HaveKeyWithValue` and compare the expected ignition content here.

File: pkg/onmetal/create_machine_test.go [hunk 2]
```diff
@@ -141,12 +137,16 @@ var _ = Describe("CreateMachine", func() {
...
 		))
 
+		ignitionJSONData := ignition.Data["ignition.json"]
```
Comment (afritzler) [inline pkg/onmetal/create_machine_test.go:145]: Move this part up as you need this later.

File: pkg/onmetal/create_machine_test.go [hunk 3]
```diff
@@ -141,12 +137,16 @@ var _ = Describe("CreateMachine", func() {
...
+		ignitionMap := make(map[string]interface{})
```
Comment (afritzler) [inline pkg/onmetal/create_machine_test.go:146]: Please call this var `ignitionData`.

File: pkg/testdata/testdata.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 OnMetal authors
```
Comment (afritzler) [inline pkg/testdata/testdata.go:1]: Please move this file to `pkg/onmetal/testing/testing.go` as this content won't be used outside.

File: pkg/testdata/testdata.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+      name: xyz
+      sshAuthorizedKeys: ssh-ed25519 AAABC3NzaC1lZDI1NTE5AAAAIGqrmrq1XwWnPJoSsAeuVcDQNqA5XQK
+      shell: /bin/bash`,
+		"dnsServers": []net.IP{
+			net.ParseIP("1.2.3.4"),
+			net.ParseIP("5.6.7.8"),
+		},
+	}
+
+	ProviderSpecWithInvalidDNS = map[string]interface{}{
```
Comment (afritzler) [inline pkg/testdata/testdata.go:50]: We don't need this - see comment below:

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/onmetal/create_machine_test.go (hunk 1 above)
Comment (sujeet01) [reply pkg/onmetal/create_machine_test.go]: Done

File: pkg/onmetal/create_machine_test.go (hunk 3 above)
Comment (sujeet01) [reply pkg/onmetal/create_machine_test.go]: Done

File: pkg/onmetal/create_machine_test.go (hunk 2 above)
Comment (sujeet01) [reply pkg/onmetal/create_machine_test.go]: Done

File: pkg/testdata/testdata.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		"ignition": `passwd:
+  users:
+    - groups: [group1]
+      name: xyz
+      sshAuthorizedKeys: ssh-ed25519 AAABC3NzaC1lZDI1NTE5AAAAIGqrmrq1XwWnPJoSsAeuVcDQNqA5XQK
+      shell: /bin/bash`,
+		"dnsServers": []net.IP{
+			net.ParseIP("1.2.3.4"),
+			net.ParseIP("5.6.7.8"),
+			net.ParseIP("invalid-dns-ips"),
```
[the remark this answers is not in the document]
Comment (sujeet01) [reply pkg/testdata/testdata.go]: Done

File: pkg/testdata/testdata.go (hunk 4 above)
Comment (sujeet01) [reply pkg/testdata/testdata.go]: Done

File: pkg/onmetal/create_machine_test.go [hunk 7]
```diff
@@ -157,20 +155,24 @@ var _ = Describe("CreateMachine", func() {
...
-				MachineClass: newMachineClass("foo", internal.ProviderSpec),
+				MachineClass: newMachineClass("foo", testing.SampleProviderSpec),
 				Secret:       providerSecret,
 			})
 			g.Expect(err.Error()).To(ContainSubstring("not supported by the driver"))
 		}).Should(Succeed())
 
 		By("failing if the invalid IP is set for dnsServers in provider")
 		Eventually(func(g Gomega) {
+			testing.SampleProviderSpec["dnsServers"] = append(testing.SampleProviderSpec["dnsServers"].([]net.IP), net.ParseIP("invalid-dns-ips"))
```
Comment (afritzler) [inline pkg/onmetal/create_machine_test.go:166]: Can you create a local copy of the `SampleProviderSpec`. That way you don't need the defer call. One option would be to implement a getter function into the `testing` package.

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/onmetal/create_machine_test.go (hunk 7 above)
Comment (sujeet01) [reply pkg/onmetal/create_machine_test.go]: Done

Comment (sujeet01) [reply pkg/onmetal/create_machine_test.go]: Done

File: pkg/onmetal/testing/testing.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+					"enabled": true,
+					"name":    "cloud-config-init.service",
+				},
+			},
+		},
+	}
+)
+
+// GetSampleProviderSpec returns a copy of the SampleProviderSpec map
+func GetSampleProviderSpec() map[string]interface{} {
```
Comment (afritzler) [inline pkg/onmetal/testing/testing.go:120]: Please create a generic map copy function here. e.g. `CopyMap(m map[string]interface{})`

File: pkg/onmetal/testing/testing.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+	newProviderSpec := make(map[string]interface{})
```
Comment (afritzler) [inline pkg/onmetal/testing/testing.go:121]:
```suggestion
	mc := make(map[string]interface{}, len(m))
```

**Review by afritzler — CHANGES_REQUESTED** [review]

File: pkg/onmetal/testing/testing.go (hunk 8 above)
Comment (sujeet01) [reply pkg/onmetal/testing/testing.go]: Done

File: pkg/onmetal/testing/testing.go (hunk 9 above)
Comment (sujeet01) [reply pkg/onmetal/testing/testing.go]: Done

**Review by afritzler — APPROVED** [review]

**Pull request merged**